fendermint/eth/api/src/cache.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::sync::{Arc, Mutex};
+
+use anyhow::Context;
+use fendermint_rpc::client::FendermintClient;
+use fendermint_rpc::query::QueryClient;
+use fvm_shared::{
+    address::{Address, Payload},
+    ActorID,
+};
+use lru_time_cache::LruCache;
+use tendermint_rpc::Client;
+
+/// Facilitate Ethereum address <-> Actor ID lookups.
+#[derive(Clone)]
+pub struct AddressCache<C> {
+    client: FendermintClient<C>,
+    addr_to_id: Arc<Mutex<LruCache<Address, ActorID>>>,
+    id_to_addr: Arc<Mutex<LruCache<ActorID, Address>>>,
+}
+
+impl<C> AddressCache<C>
+where
+    C: Client + Sync + Send,
+{
+    pub fn new(client: FendermintClient<C>, capacity: usize) -> Self {
+        Self {
+            client,
+            addr_to_id: Arc::new(Mutex::new(LruCache::with_capacity(capacity))),
+            id_to_addr: Arc::new(Mutex::new(LruCache::with_capacity(capacity))),
+        }
+    }
+
+    pub async fn lookup_id(&self, addr: &Address) -> anyhow::Result<Option<ActorID>> {
+        if let Ok(id) = addr.id() {
+            return Ok(Some(id));
+        }
+
+        if let Some(id) = self.get_id(addr) {
+            return Ok(Some(id));
+        }
+
+        let res = self
+            .client
+            .actor_state(addr, None)
+            .await
+            .context("failed to lookup actor state")?;
+
+        match res.value {
+            Some((id, _)) => {
+                self.set_id(*addr, id);
+                if let Payload::Delegated(_) = addr.payload() {
```

Comment:
Why do you need to explicitly check if it is a delegated address to index the id by address, and you don't check to index by address by id?

Reply:
When I look up the `Address` by ID, all I can get is the optional `delegated_address` field in the `ActorState`. However if you look up an ID by `Address`, the system can handle any type of address, so it's not symmetric. I only want to insert into the id-to-address if it's the same kind of address as what I'd do if I did the opposite lookup. I don't want to fill in the space with something that isn't a delegated address. It's contrived, I know.